File: 72414.cpp
#include <string>
#include <vector>
#include <stdio.h>
#include <queue>

using namespace std;

int timeToSeconds(string time) {
    int hours = stoi(string(time.begin(), time.begin() + 2));
    int minutes = stoi(string(time.begin() + 3, time.begin() + 5));
    int seconds = stoi(string(time.begin() + 6, time.begin() + 8));
    
    return hours * 3600 + minutes * 60 + seconds;
}

void logToTimes(vector<int>& seconds, string log) {
    int from = timeToSeconds(string(log.begin(), log.begin() + 8));
    int to = timeToSeconds(string(log.begin() + 9, log.end()));
    
    while(from < to) {
        seconds[from] += 1;
        from++;
    }
}

string timeToString(int time) {
    int hours = time / 3600;
    int minutes = (time - hours * 3600) / 60;
    int seconds = time - (hours * 3600) - (minutes * 60);
    
    char tempString [9];
    sprintf(tempString, "%02d:%02d:%02d", hours, minutes, seconds);
    
    return string(tempString);
}

string solution(string play_time, string adv_time, vector<string> logs) {
    string answer = "";
    int endTime = timeToSeconds(play_time);
    int advTime = timeToSeconds(adv_time);
    vector<int> seconds(endTime, 0);
    
    for(string log : logs) {
        logToTimes(seconds, log);
    }
    
    long long bestSum = 0;
    int bestFrom = 0;
    
    long long sum = 0;
    int from = 0;
    int to = advTime;
    
    queue<int> searchQueue;
    
    for(int i = 0; i <= to; i++) {
        searchQueue.push(seconds[i]);
        sum += seconds[i];
    }
    bestSum = sum;
    
    while(to <= endTime) {
        sum -= searchQueue.front();
        sum += seconds[to];
        
        from++;
        searchQueue.pop();
        to++;
        searchQueue.push(seconds[to]);
        
        if(sum > bestSum) {
            bestSum = sum;
            bestFrom = from;
        }
    }
    
    return timeToString(bestFrom);
}

File: 72414.swift
import Foundation

func toSeconds(_ time: String) -> Int {
    let splited = time.components(separatedBy: ":")
    guard let hours = Int(splited[0]), let minutes = Int(splited[1]), let seconds = Int(splited[2]) else { return 0 }
    return hours * 3600 + minutes * 60 + seconds
}

func toString(_ time: Int) -> String {
    let hours = time / 3600
    let minutes = (time - hours * 3600) / 60
    let seconds = time - hours * 3600 - minutes * 60
    return String(format: "%02d:%02d:%02d", hours < 0 ? 0 : hours, minutes < 0 ? 0 : minutes, seconds < 0 ? 0 : seconds)
}

func solution(_ play_time:String, _ adv_time:String, _ logs:[String]) -> String {
    let playTime = toSeconds(play_time)
    var adTime = toSeconds(adv_time)
    var accumulatedViews = Array(repeating: 0, count: playTime + 1)

    for log in logs {
        let splited = log.components(separatedBy: "-")
        accumulatedViews[toSeconds(splited[0])] += 1
        accumulatedViews[toSeconds(splited[1])] -= 1
    }
    
    for i in 1..<playTime { accumulatedViews[i] += accumulatedViews[i - 1] }
    for i in 1..<playTime { accumulatedViews[i] += accumulatedViews[i - 1] }
    
    var from = 0
    var to = adTime

    var bestViews = 0
    var bestTime = 0
    
    while to <= playTime {
        let currentViews = accumulatedViews[to - 1] - (from - 1 > 0 ? accumulatedViews[from - 1] : 0)
        if currentViews > bestViews {
            bestViews = currentViews
            bestTime = from
        }
        
        from += 1
        to += 1
    }
    
    return toString(bestTime)
}
